### PR TITLE
[SC] Contract balance query APIs for parent/child bridge contracts

### PR DESCRIPTION
**Pull request by hyunsooda**
## Proposed changes

The bridge contract deployed and registered must have enough balance than the requested amount of KLAY. If not, the counterpart chain's receiver does not actually receive that amount of KLAY, but the sender's KLAY was already transferred to the bridge contract and not refunded back. This PR suggests a new external API to be used by end-developers so that they can use this API first to get the bridge contract's balance on the counterpart chain. 

TODO (To-be-discussed):
 - The current implementation of the new API takes an address parameter. In my books, a new individual bridge management structure seems needed. For example, ServiceChain (SC) network can have numerous bridge pairs in theory, but the current implementation does not have a management context, wherein the end-developers must keep the bridge address themselves to interact with proper bridge pairs. Thus, a feature of pair name context seems needed.  

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: node/sc/bridge_manager_test.go [hunk 1]
```diff
@@ -131,17 +132,17 @@ func TestBridgeManager(t *testing.T) {
 		return
 	}
 
-	bridgeManager, err := NewBridgeManager(sc)
+	bm, err := NewBridgeManager(sc)
```
Comment (kjhman21) [inline node/sc/bridge_manager_test.go:135]: @hyunsooda Please do not make unnecessary changes. You can make a separate PR for this name change.

Comment (hyunsooda) [reply node/sc/bridge_manager_test.go]: Reverted. Let me post another PR for reverted one later.

File: node/sc/api_bridge.go [hunk 2]
```diff
@@ -510,10 +511,19 @@ func (sb *SubBridgeAPI) GetParentOperatorBalance() (*big.Int, error) {
 	return sb.subBridge.handler.getParentOperatorBalance()
 }
 
+// GetParentBridgeContractBalance returns bridge contract's balance from the parent bridge
```
Comment (kjhman21) [inline node/sc/api_bridge.go:514]:
```suggestion
// GetParentBridgeContractBalance returns the balance of the bridge contract in the parent chain.
```

File: node/sc/api_bridge.go [hunk 3]
```diff
@@ -510,10 +511,19 @@ func (sb *SubBridgeAPI) GetParentOperatorBalance() (*big.Int, error) {
...
+func (sb *SubBridgeAPI) GetParentBridgeContractBalance(addr common.Address) (*big.Int, error) {
+	return sb.subBridge.handler.getParentBridgeContractBalance(addr)
+}
+
 func (sb *SubBridgeAPI) GetChildOperatorBalance() (*big.Int, error) {
 	return sb.subBridge.handler.getChildOperatorBalance()
 }
 
+func (sb *SubBridgeAPI) GetChildBridgeContractBalance(addr common.Address) (*big.Int, error) {
```
Comment (kjhman21) [inline node/sc/api_bridge.go:523]:
```suggestion
// GetChildBridgeContractBalance returns the balance of the bridge contract in the child chain.
func (sb *SubBridgeAPI) GetChildBridgeContractBalance(addr common.Address) (*big.Int, error) {
```

File: node/sc/bridge_manager.go [hunk 4]
```diff
@@ -588,6 +588,24 @@ func (bm *BridgeManager) GetAllBridge() []*BridgeJournal {
 	return gwjs
 }
 
+// GetAllChildBridgeAddrs returns a slice of child bridge address list.
```
Comment (kjhman21) [inline node/sc/bridge_manager.go:591]:
```suggestion
// GetAllChildBridgeAddrs returns a list of child bridge contract addresses.
```

File: node/sc/bridge_manager.go [hunk 5]
```diff
@@ -588,6 +588,24 @@ func (bm *BridgeManager) GetAllBridge() []*BridgeJournal {
...
+func (bm *BridgeManager) GetAllChildBridgeAddrs() []common.Address {
+	var childAddrs []common.Address
+	for _, journal := range bm.journal.cache {
+		childAddrs = append(childAddrs, journal.ChildAddress)
+	}
+	return childAddrs
+}
+
+// GetAllParentBridgeAddrs returns a slice of parent bridge address list.
```
Comment (kjhman21) [inline node/sc/bridge_manager.go:600]:
```suggestion
// GetAllParentBridgeAddrs returns a list of parent bridge contract addresses.
```

File: node/sc/bridge_manager_test.go [hunk 6]
```diff
@@ -1988,7 +1989,7 @@ func TestDecodingLegacyAnchoringTx(t *testing.T) {
 }
 
 // DeployBridgeTest is a test-only function which deploys a bridge contract with some amount of KLAY.
-func (bm *BridgeManager) DeployBridgeTest(backend *backends.SimulatedBackend, local bool) (common.Address, error) {
+func (bm *BridgeManager) DeployBridgeTest(backend *backends.SimulatedBackend, amountOfBridgeContract int64, local bool) (common.Address, error) {
```
Comment (kjhman21) [inline node/sc/bridge_manager_test.go:1992]: `amountOfBridgeContract` is for depositing KLAY to the bridge contract, right?

What about...? I don't know what is good...
- KLAYinPeb
- depositKLAYinPeb
- value

File: node/sc/sub_bridge_handler.go [hunk 7]
```diff
@@ -122,6 +123,25 @@ func (sbh *SubBridgeHandler) getParentOperatorBalance() (*big.Int, error) {
 	return sbh.subbridge.remoteBackend.BalanceAt(ctx, sbh.subbridge.bridgeAccounts.pAccount.address, nil)
 }
 
+// getParentBridgeContractBalance returns the parent bridge contract's balance.
+func (sbh *SubBridgeHandler) getParentBridgeContractBalance(addr common.Address) (*big.Int, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	found := false
+	for _, bridgeContractAddr := range sbh.subbridge.bridgeManager.GetAllParentBridgeAddrs() {
```
Comment (kjhman21) [inline node/sc/sub_bridge_handler.go:132]: What do you think?

1. merge this first and later implement alias.
2. develop this code base with alias

File: node/sc/bridge_manager_test.go (hunk 6 above)
Comment (hyunsooda) [reply node/sc/bridge_manager_test.go:1992]: Yes. The list-up seems too general. I think it would be more understandable if the variable name contains the semantic of that value. So, how about `amountOfDeposit`?

File: node/sc/sub_bridge_handler.go (hunk 7 above)
Comment (hyunsooda) [reply node/sc/sub_bridge_handler.go:132]: I'm okay with anything. but the separate PR would be nice in view of reviewers.

File: node/sc/bridge_manager_test.go (hunk 6 above)
Comment (kjhman21) [reply node/sc/bridge_manager_test.go:1992]: @hyunsooda Okay. I'm good with `amountOfDeposit`.

File: node/sc/sub_bridge_handler.go (hunk 7 above)
Comment (kjhman21) [reply node/sc/sub_bridge_handler.go:132]: Okay, then let's go with the first. Thanks.

File: node/sc/bridge_manager_test.go (hunk 6 above)
Comment (hyunsooda) [reply node/sc/bridge_manager_test.go:1992]: Renamed.

File: node/sc/sub_bridge_handler.go [hunk 8]
```diff
@@ -138,6 +158,25 @@ func (sbh *SubBridgeHandler) getChildOperatorBalance() (*big.Int, error) {
 	return sbh.subbridge.localBackend.BalanceAt(ctx, sbh.subbridge.bridgeAccounts.cAccount.address, nil)
 }
 
+// getChildBridgeContractBalance returns the child bridge contract's balance.
+func (sbh *SubBridgeHandler) getChildBridgeContractBalance(addr common.Address) (*big.Int, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	found := false
+	for _, bridgeContractAddr := range sbh.subbridge.bridgeManager.GetAllChildBridgeAddrs() {
```
Comment (kjhman21) [inline node/sc/sub_bridge_handler.go:167]: Do we really make a slice for this function? What about adding address check logic into `bridgeManager`?
Then, we can call like `bridgeManager.IsInChildBridge(addr)`?

Comment (hyunsooda) [reply node/sc/sub_bridge_handler.go]: Refactored. PTAL.

File: node/sc/bridge_manager.go [hunk 9]
```diff
@@ -1143,3 +1161,23 @@ func (bm *BridgeManager) GetFeeReceiver(bridgeAddr common.Address) (common.Addre
 
 	return bi.bridge.FeeReceiver(nil)
 }
+
+// IsInParentAddrs returns true if the bridgeAddr is in the list of parent bridge addresses and returns false if not.
+func (bm *BridgeManager) IsInParentAddrs(bridgeAddr common.Address) bool {
+	return bm.IsValidBridgeAddr(bridgeAddr, bm.GetAllParentBridgeAddrs())
```
Comment (kjhman21) [inline node/sc/bridge_manager.go:1167]:
```suggestion
	for _, journal := range bm.journal.cache {
		if journal.ChildAddress == bridgeAddr {
			return true
		}
	}
	return false
```

something like this?

Comment (hyunsooda) [reply node/sc/bridge_manager.go:1167]: 1. Okay, then how about making it as a function? 
2. The functions `GetAllChildbridgeAddrs()` and `GetAllParentBridgeAddrs` are now not called if your change is reflected. I think leaving them as seems okay. What do you think?

Comment (kjhman21) [reply node/sc/bridge_manager.go:1167]: 1. I think it is already a function?
2. yes. removing them is good to me.

**Review by KimKyungup — DISMISSED** [review]
New APIs are good.
It is better that if we add more APIs for FT,NFT.

Please check the linter.

**Comment by hyunsooda** [comment]
Hi @KimKyungup,

Thanks for the suggestion. An issue is posted at #1296.

File: node/sc/bridge_manager.go (hunk 9 above)
Comment (hyunsooda) [reply node/sc/bridge_manager.go]: I'm sorry. I mixed up that logic is in `sub_bridge_handler.go`. Updated with a simplified one.

**Review by aidan-kwon — DISMISSED** [review]

**Review by KimKyungup — APPROVED** [review]

**Comment by hyunsooda** [comment]
@aidan-kwon, Resolved.

**Review by aidan-kwon — APPROVED** [review]

**Comment by aidan-kwon** [comment]
Reviewers, PTAL

**Review by kjhman21 — APPROVED** [review]

**Review by hqjang-pepper — APPROVED** [review]

**Review by kjhman21 — APPROVED** [review]
